File: flutter/android/src/main/kotlin/com/pylon/chatwidget/flutter/PylonChatPlatformView.kt
package com.pylon.chatwidget.flutter

import android.content.Context
import android.view.View
import com.pylon.chatwidget.PylonChat
import com.pylon.chatwidget.PylonChatListener
import com.pylon.chatwidget.PylonConfig
import com.pylon.chatwidget.PylonUser
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/**
 * Hosts a [PylonChat] view inside Flutter and bridges it to Dart.
 *
 * Method calls arrive from `PylonChatController`; widget events go back out on
 * the same channel, where `PylonChatView` turns them into callbacks and into
 * the hit test regions that let touches pass through to the Flutter app behind
 * the widget.
 */
class PylonChatPlatformView(
    context: Context,
    messenger: BinaryMessenger,
    viewId: Int,
    creationParams: Map<*, *>
) : PlatformView, MethodChannel.MethodCallHandler, PylonChatListener {

    private val channel = MethodChannel(messenger, "$CHANNEL_PREFIX$viewId")

    /** Web view bounds arrive in device pixels; Flutter lays out in logical pixels. */
    private val density = context.resources.displayMetrics.density

    private val chatView: PylonChat
    private var currentUser: PylonUser?
    private var isChatOpen = false

    init {
        currentUser = parseUser(creationParams["user"])
        chatView = PylonChat(context, parseConfig(creationParams["config"]), currentUser)
        chatView.setListener(this)
        channel.setMethodCallHandler(this)
    }

    override fun getView(): View = chatView

    override fun dispose() {
        channel.setMethodCallHandler(null)
        chatView.setListener(null)
        chatView.destroy()
    }

    // MARK: - Dart → native

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "openChat" -> chatView.openChat()
            "closeChat" -> chatView.closeChat()
            "showChatBubble" -> chatView.showChatBubble()
            "hideChatBubble" -> chatView.hideChatBubble()
            "refreshInteractiveBounds" -> chatView.refreshInteractiveBounds()

            "showNewMessage" -> chatView.showNewMessage(
                call.argument<String>("message").orEmpty(),
                call.argument<Boolean>("isHtml") ?: false
            )

            "setNewIssueCustomFields" -> chatView.setNewIssueCustomFields(
                stringKeyedMap(call.argument<Map<*, *>>("fields"))
            )

            "setTicketFormFields" -> chatView.setTicketFormFields(
                stringKeyedMap(call.argument<Map<*, *>>("fields"))
            )

            "showTicketForm" -> chatView.showTicketForm(
                call.argument<String>("slug").orEmpty()
            )

            "showKnowledgeBaseArticle" -> chatView.showKnowledgeBaseArticle(
                call.argument<String>("articleId").orEmpty()
            )

            "updateUser" -> {
                currentUser = parseUser(call.argument<Map<*, *>>("user"))
                chatView.setUser(currentUser)
            }

            "updateEmailHash" -> {
                val user = currentUser
                if (user == null) {
                    result.error(
                        "no_user",
                        "Set a user before setting an email hash.",
                        null
                    )
                    return
                }
                currentUser = user.copy(emailHash = call.argument<String>("emailHash"))
                chatView.setUser(currentUser)
            }

            else -> {
                result.notImplemented()
                return
            }
        }
        result.success(null)
    }

    // MARK: - Native → Dart

    override fun onPylonLoaded() = send("onPylonLoaded")

    override fun onPylonInitialized() = send("onPylonInitialized")

    override fun onPylonReady() = send("onPylonReady")

    override fun onMessageReceived(message: String) =
        send("onMessageReceived", mapOf("message" to message))

    override fun onChatOpened() {
        isChatOpen = true
        send("onChatOpened")
    }

    override fun onChatClosed() {
        val wasOpen = isChatOpen
        isChatOpen = false
        send("onChatClosed", mapOf("wasOpen" to wasOpen))
    }

    override fun onPylonError(error: String) =
        send("onPylonError", mapOf("error" to error))

    override fun onUnreadCountChanged(unreadCount: Int) =
        send("onUnreadCountChanged", mapOf("count" to unreadCount))

    override fun onInteractiveBoundsChanged(
        selector: String,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float
    ) = send(
        // Only a scale is needed here: the WebView lays its content out from its own
        // origin, so the page's coordinates and the view's already share a space.
        // (iOS differs — WKWebView insets its content for the safe area, so the
        // bridge there has to shift the bounds as well.)
        "onInteractiveBoundsChanged",
        mapOf(
            "selector" to selector,
            "left" to (left / density).toDouble(),
            "top" to (top / density).toDouble(),
            "right" to (right / density).toDouble(),
            "bottom" to (bottom / density).toDouble()
        )
    )

    private fun send(method: String, arguments: Map<String, Any?>? = null) {
        channel.invokeMethod(method, arguments)
    }

    // MARK: - Creation params

    private fun parseConfig(raw: Any?): PylonConfig {
        val map = raw as? Map<*, *> ?: error("PylonChatView was created without a config.")
        val appId = map["appId"] as? String
            ?: error("PylonChatView was created without an appId.")

        return PylonConfig.build(appId) {
            enableLogging = map["enableLogging"] as? Boolean ?: true
            debugMode = map["debugMode"] as? Boolean ?: false
            primaryColor = map["primaryColor"] as? String
            (map["widgetBaseUrl"] as? String)?.let { widgetBaseUrl = it }
            (map["widgetScriptUrl"] as? String)?.let { widgetScriptUrl = it }
        }
    }

    private fun parseUser(raw: Any?): PylonUser? {
        val map = raw as? Map<*, *> ?: return null
        val email = map["email"] as? String ?: return null
        val name = map["name"] as? String ?: return null
        return PylonUser(
            email = email,
            name = name,
            avatarUrl = map["avatarUrl"] as? String,
            emailHash = map["emailHash"] as? String,
            accountId = map["accountId"] as? String,
            accountExternalId = map["accountExternalId"] as? String
        )
    }

    private fun stringKeyedMap(raw: Map<*, *>?): Map<String, Any?> {
        if (raw == null) return emptyMap()
        return raw.entries
            .mapNotNull { (key, value) -> (key as? String)?.let { it to value } }
            .toMap()
    }

    companion object {
        /** Must match `_PylonChatViewState._channelPrefix` on the Dart side. */
        private const val CHANNEL_PREFIX = "com.pylon.chatwidget/pylon_chat_view_"
    }
}

File: flutter/ios/Classes/PylonChatPlatformView.swift
import Flutter
import UIKit

/// Hosts a `PylonChatView` inside Flutter and bridges it to Dart.
///
/// Method calls arrive from `PylonChatController`; widget events go back out on
/// the same channel, where `PylonChatView` turns them into callbacks and into
/// the hit test regions that let touches pass through to the Flutter app behind
/// the widget.
class PylonChatPlatformView: NSObject, FlutterPlatformView {

    private let channel: FlutterMethodChannel
    private let container: PylonChatContainerView

    /// Calls that arrived before the SDK view had been laid out into existence.
    private var pendingCalls: [(FlutterMethodCall, FlutterResult)] = []

    init(frame: CGRect,
         viewId: Int64,
         arguments: Any?,
         messenger: FlutterBinaryMessenger) {
        let params = arguments as? [String: Any] ?? [:]

        channel = FlutterMethodChannel(
            name: "\(PylonChatPlatformView.channelPrefix)\(viewId)",
            binaryMessenger: messenger
        )
        container = PylonChatContainerView(
            config: PylonChatPlatformView.parseConfig(params["config"]),
            user: PylonChatPlatformView.parseUser(params["user"]),
            frame: frame
        )

        super.init()

        container.listener = self
        container.onChatViewCreated = { [weak self] in
            self?.flushPendingCalls()
        }
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }
    }

    private func flushPendingCalls() {
        let queued = pendingCalls
        pendingCalls = []
        for (call, result) in queued {
            handle(call, result: result)
        }
    }

    func view() -> UIView {
        return container
    }

    deinit {
        channel.setMethodCallHandler(nil)
        // Anything still queued will never run; fail it rather than leaving the
        // Dart side awaiting a reply that cannot come.
        for (_, result) in pendingCalls {
            result(FlutterError(
                code: "disposed",
                message: "The Pylon chat widget was disposed.",
                details: nil
            ))
        }
        pendingCalls = []
        container.destroy()
    }

    // MARK: - Dart → native

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let chatView = container.chatView else {
            // The SDK view is built on the first layout pass with a real size, which
            // can land after the controller has already flushed its buffered calls.
            // Hold them until there is something to act on.
            pendingCalls.append((call, result))
            return
        }

        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "openChat":
            chatView.openChat()
        case "closeChat":
            chatView.closeChat()
        case "showChatBubble":
            chatView.showChatBubble()
        case "hideChatBubble":
            chatView.hideChatBubble()
        case "refreshInteractiveBounds":
            chatView.refreshInteractiveBounds()
        case "showNewMessage":
            chatView.showNewMessage(
                args["message"] as? String ?? "",
                isHtml: args["isHtml"] as? Bool ?? false
            )
        case "setNewIssueCustomFields":
            chatView.setNewIssueCustomFields(Self.fields(from: args["fields"]))
        case "setTicketFormFields":
            chatView.setTicketFormFields(Self.fields(from: args["fields"]))
        case "showTicketForm":
            chatView.showTicketForm(args["slug"] as? String ?? "")
        case "showKnowledgeBaseArticle":
            chatView.showKnowledgeBaseArticle(args["articleId"] as? String ?? "")
        case "updateUser":
            guard let user = Self.parseUser(args["user"]) else {
                result(FlutterError(
                    code: "invalid_user",
                    message: "A Pylon user needs both an email and a name.",
                    details: nil
                ))
                return
            }
            container.updateUser(user)
        case "updateEmailHash":
            guard let user = container.currentUser else {
                result(FlutterError(
                    code: "no_user",
                    message: "Set a user before setting an email hash.",
                    details: nil
                ))
                return
            }
            container.updateUser(PylonUser(
                email: user.email,
                name: user.name,
                avatarUrl: user.avatarUrl,
                emailHash: args["emailHash"] as? String,
                accountId: user.accountId,
                accountExternalId: user.accountExternalId
            ))
        default:
            result(FlutterMethodNotImplemented)
            return
        }

        result(nil)
    }

    // MARK: - Creation params

    private static func parseConfig(_ raw: Any?) -> PylonConfig {
        let map = raw as? [String: Any] ?? [:]
        return PylonConfig(
            appId: map["appId"] as? String ?? "",
            enableLogging: map["enableLogging"] as? Bool ?? true,
            primaryColor: map["primaryColor"] as? String,
            debugMode: map["debugMode"] as? Bool ?? false,
            widgetBaseUrl: map["widgetBaseUrl"] as? String,
            widgetScriptUrl: map["widgetScriptUrl"] as? String,
            // Points map 1:1 onto Flutter's logical pixels, so no scaling needed.
            bubbleBottomOffset: (map["bubbleBottomOffset"] as? NSNumber)?.doubleValue ?? 0
        )
    }

    private static func parseUser(_ raw: Any?) -> PylonUser? {
        guard let map = raw as? [String: Any],
              let email = map["email"] as? String,
              let name = map["name"] as? String else {
            return nil
        }
        return PylonUser(
            email: email,
            name: name,
            avatarUrl: map["avatarUrl"] as? String,
            emailHash: map["emailHash"] as? String,
            accountId: map["accountId"] as? String,
            accountExternalId: map["accountExternalId"] as? String
        )
    }

    private static func fields(from raw: Any?) -> [String: Any] {
        guard let map = raw as? [String: Any] else { return [:] }
        return map.filter { !($0.value is NSNull) }
    }

    /// Must match `_PylonChatViewState._channelPrefix` on the Dart side.
    private static let channelPrefix = "com.pylon.chatwidget/pylon_chat_view_"
}

// MARK: - Native → Dart

extension PylonChatPlatformView: PylonChatListener {

    func onPylonLoaded() {
        channel.invokeMethod("onPylonLoaded", arguments: nil)
    }

    func onPylonInitialized() {
        channel.invokeMethod("onPylonInitialized", arguments: nil)
    }

    func onPylonReady() {
        channel.invokeMethod("onPylonReady", arguments: nil)
    }

    func onMessageReceived(message: String) {
        channel.invokeMethod("onMessageReceived", arguments: ["message": message])
    }

    func onChatOpened() {
        channel.invokeMethod("onChatOpened", arguments: nil)
    }

    func onChatClosed(wasOpen: Bool) {
        channel.invokeMethod("onChatClosed", arguments: ["wasOpen": wasOpen])
    }

    func onPylonError(error: String) {
        channel.invokeMethod("onPylonError", arguments: ["error": error])
    }

    func onUnreadCountChanged(count: Int) {
        channel.invokeMethod("onUnreadCountChanged", arguments: ["count": count])
    }

    func onInteractiveBoundsChanged(selector: String, bounds: CGRect) {
        // Points map 1:1 onto Flutter's logical pixels, so no scaling — but the
        // page's origin sits inside the view's by the web view's content inset,
        // and Flutter hit tests in view coordinates.
        let offset = container.contentOffset
        channel.invokeMethod("onInteractiveBoundsChanged", arguments: [
            "selector": selector,
            "left": bounds.minX + offset.x,
            "top": bounds.minY + offset.y,
            "right": bounds.maxX + offset.x,
            "bottom": bounds.maxY + offset.y
        ])
    }
}
